src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -520,14 +520,81 @@ void MPS::apply_matrix(const reg_t & qubits, const cmatrix_t &mat)
   case 1: 
     q_reg_[qubits[0]].apply_matrix(mat);
     break;
-  case 2:
-    apply_2_qubit_gate(qubits[0], qubits[1], su4, mat);
-    break;
+    //  case 2:
+    //    apply_2_qubit_gate(qubits[0], qubits[1], su4, mat);
+    //    break;
   default:
-    throw std::invalid_argument("currently support apply_matrix for 1 or 2 qubits only");
+    apply_multi_qubit_gate(qubits, mat);
   }
 }
 
+void MPS::apply_multi_qubit_gate(const reg_t &qubits,
+				 const cmatrix_t &mat) {
+  uint_t num_qubits = qubits.size();
+  // need to reverse qubits because that is the way they
+  // are defined in the Qiskit interface
+  reg_t reversed_qubits = qubits;
+  std::reverse(reversed_qubits.begin(), reversed_qubits.end()); 
+
+  bool ordered = true;
+  for (uint_t index=0; index < qubits.size()-1; index++) {
+    if (reversed_qubits[index]+1 != reversed_qubits[index+1]){
+      ordered = false;
+      break;
+    }
+  }
+  reg_t actual_indices(num_qubits_);
+  std::iota( std::begin(actual_indices), std::end(actual_indices), 0);
+  reg_t target_qubits(num_qubits);
+  if (ordered) {
+    target_qubits = reversed_qubits;
+  } else {
+    move_qubits_to_right_end(reversed_qubits, target_qubits, actual_indices);
+  }
+
+  uint_t first = target_qubits.front();
+  MPS_Tensor sub_tensor(state_vec_as_MPS(first, first+num_qubits-1));
+
+  sub_tensor.apply_matrix(mat, false /* swapped */);
+
+
+  // state_mat is a matrix containing the flattened representation of the sub-tensor 
+  // into a single matrix. Note that sub_tensor will contain 8 matrices for 3-qubit
+  // gates. state_mat will be the concatenation of them all.
+  cmatrix_t state_mat = sub_tensor.get_data(0);
+  for (uint_t i=1; i<sub_tensor.get_data().size(); i++)
+    state_mat = AER::Utils::concatenate(state_mat, sub_tensor.get_data(i), 1) ;
+
+  // We convert the matrix back into an MPS structure
+  MPS sub_MPS;
+  sub_MPS.initialize_from_matrix(qubits.size(), state_mat);
+
+  if (num_qubits == num_qubits_) {
+    q_reg_.clear();
+    q_reg_ = sub_MPS.q_reg_;
+    lambda_reg_ = sub_MPS.lambda_reg_;
+  } else {
+    // copy the sub_MPS back to the corresponding positions in the original MPS
+    for (uint_t i=0; i<sub_MPS.num_qubits(); i++) {
+      q_reg_[first+i] = sub_MPS.q_reg_[i];
+    }
+    lambda_reg_[first] = sub_MPS.lambda_reg_[0];
+    if (first > 0)
+      q_reg_[first].div_Gamma_by_left_Lambda(lambda_reg_[first-1]);
+
+    for (uint_t i=1; i<num_qubits-1; i++) {
+      lambda_reg_[first+i] = sub_MPS.lambda_reg_[i]; 
+    }
+    if (first+num_qubits-1 < num_qubits_-1)
+	q_reg_[first+num_qubits-1].div_Gamma_by_right_Lambda(lambda_reg_[first+num_qubits-1]);
+      
+  }
+  // need to move qubits back to original position, if they were moved
+  // at the beginning
+  if (!ordered) {
```

Comment:
Just wondering if we can avoid all this moving qubits back and forth somehow.
Maybe having two functions like: `void MPS::apply_multi_qubit_gate()` and `void MPS::apply_ordered_multi_qubit_gate()` ?